### ISSUE TITLE
How to specify array type with specDescriptors

### ISSUE THREAD

**Issue by aharbis** [openshift/console#5260]
I'm using OCP 4.3, and trying to build out our CSV for our Operator. We have a property in our CRD that is an array of strings, which represent Kubernetes Secrets. Using the following `specDescriptors`:

```go
// Image pull secrets
// +kubebuilder:validation:Optional
// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors=true
// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors.displayName="Image pull secrets"
// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors.x-descriptors="urn:alm:descriptor:io.kubernetes:Secret"
ImagePullSecrets []string `json:"imagePullSecrets"`
```

the following is rendered in the form in OLM:

![image](https://user-images.githubusercontent.com/20443310/80819778-a73dd980-8ba3-11ea-9ba6-888a2cdefef6.png)

I've reviewed the [OLM Descriptor Reference](https://github.com/openshift/console/blob/master/frontend/packages/operator-lifecycle-manager/src/components/descriptors/reference/reference.md) but I cannot figure out how to render this as a proper array, and allow users to specify multiple Secrets.

We have several fields that are arrays, so we would like to find a solution that applies to multiple types, not just `Secret`. Some other types that I would like to represent in array form are:

- `io.kubernetes:Secret`
- `io.kubernetes:ConfigMap`
- `com.tectonic.ui:text`

In addition to the above, we have complex (custom) types that are available to the user in array form.

Is it possible to declare fields as array of a given type, such that the console will allow the user to specify multiple values?

Version info for reference:

```
$ oc version
Client Version: openshift-clients-4.3.0-201910250623-88-g6a937dfe
Server Version: 4.3.0
Kubernetes Version: v1.16.2

$ operator-sdk version
operator-sdk version: "v0.17.0", commit: "2fd7019f856cdb6f6618e2c3c80d15c3c79d1b6c", kubernetes version: "unknown", go version: "go1.13.10 linux/amd64"
```

**Comment by spadgett** [comment]
cc @tlwu2013 @TheRealJon

**Comment by TheRealJon** [comment]
@aharbis Could you maybe post some YAML snippets for your CRD `spec.validation.openAPIV3Schema` and/or the matching CSV `spec.customresourcedefinitions.owned`? These are the data sources that are used to generate the form. Also, maybe some example of things you tried that didn't work as expected?

**Comment by aharbis** [comment]
Here's the full end-to-end example of our `ImagePullSecrets` property.

```go
// Image pull secrets
// +kubebuilder:validation:Optional
// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors=true
// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors.displayName="Image pull secrets"
// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors.x-descriptors="urn:alm:descriptor:io.kubernetes:Secret"
ImagePullSecrets []string `json:"imagePullSecrets"`
```

CSV:

```yaml
- description: Image pull secrets
  displayName: Image pull secrets
  path: imagePullSecrets
  x-descriptors:
  - urn:alm:descriptor:io.kubernetes:Secret
```

CRD:

```yaml
imagePullSecrets:
  description: Image pull secrets
  items:
    type: string
  type: array
```

In the console, it's rendered as a single drop-down of type `Secret`. There is no way to specify or select more than one. I have an example screenshot in my original post.

I'm not really sure what else I could "try." There's no way to tell the console via `specDescriptor` that this is an array item. It should be able to tell that via the `type` right?

**Comment by TheRealJon** [comment]
@aharbis 

If you look at the [docs](https://github.com/openshift/console/blob/release-4.3/frontend/packages/operator-lifecycle-manager/src/components/descriptors/reference/reference.md#array-field-group) for 4.3, under the Ceation View, there is a note that the widget doesn't support add/remove capabilities. That feature is added in 4.4, and will be further improved in 4.5. In 4.3, the arrayFieldGroup descriptor essentially just mimics the fieldGroup descriptor, and only supports array items of object type.

**Comment by aharbis** [comment]
Thanks @TheRealJon. I was actually noticing that in 4.4 I was seeing some of our objects render with the add/remove capability, but not all. For example, the `ImagePullSecret` example above still only provides a single drop-down option, even on 4.4. However, we have a `Users` object array, and that offered the add/remove capability.

**Comment by TheRealJon** [comment]
@aharbis 

> Thanks @TheRealJon. I was actually noticing that in 4.4 I was seeing some of our objects render with the add/remove capability, but not all. For example, the `ImagePullSecret` example above still only provides a single drop-down option, even on 4.4. 


So, 4.4 added the ability to add/remove items to an array field, but still only supports object type items. If adjusting your schema is an option, you can fix this issue by making each of your imagePullSecret items an object with a single string property:

```yaml
imagePullSecrets:
  description: Image pull secrets
  type: array
  items:
    type: object
    properties:
      secret:
        type: string
```
                 

> However, we have a `Users` object array, and that offered the add/remove capability.

My guess is that your schema property for users is something like:
```yaml
users:
  type: array
  items:
    type: object
    properties:
      # ...
``` 
In which case, it is supported because the users.items.type is 'object'.

**Comment by TheRealJon** [comment]
Note that in 4.5, array type schemas are fully supported and this will no longer be an issue. So if you don't want to change your schema, you can wait and this will just work as you are expecting in 4.5.

**Comment by spadgett** [comment]
This is fixed by #4944 in 4.5

/close
